bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/DownloadServlet.java

```diff
@@ -71,94 +81,128 @@ protected void doGet(@Nonnull SlingHttpServletRequest request, @Nonnull SlingHtt
-                    if (calendar != null) {
-                        if (checkLastModifiedAfter(calendar.getTimeInMillis(), request)) {
-                            response.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
-                            return;
-                        }
-
-                        String mimeType = valueMap.get(JcrConstants.JCR_MIMETYPE, String.class);
-                        if (StringUtils.isNotEmpty(mimeType)) {
-                            sendResponse(bytes, mimeType, filename, calendar.getTimeInMillis(), response, checkForInlineSelector(request));
-                        } else {
-                            response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                        }
-
-                    } else {
-                        response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                    }
-                }
+                sendResource(request, response, filename, downloadDataResource);
             }
         } else {
-            byte[] bytes = getByteArray(asset);
-            if (bytes.length == 0) {
-                response.sendError(HttpServletResponse.SC_NOT_FOUND);
-                return;
-            }
-
+            
             if (checkLastModifiedAfter(asset.getLastModified(), request)) {
                 response.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
                 return;
             }
-            sendResponse(bytes, asset.getMimeType(), asset.getName(), asset.getLastModified(), response, checkForInlineSelector(request));
+            Optional<InputStream> inputStream = null;
+            try {
+                inputStream = getInputStream(asset);
+                if (!inputStream.isPresent()) {
+                    response.sendError(HttpServletResponse.SC_NOT_FOUND);
+                    return;
+                }
+                sendResponse(inputStream.get(), asset.getOriginal().getSize(), asset.getMimeType(), asset.getName(), asset.getLastModified(), response, checkForInlineSelector(request));
+            } finally {
+                if (inputStream != null && inputStream.isPresent()) {
+                    inputStream.get().close();
+                }
+            }
         }
     }
 
-    private boolean checkForInlineSelector(SlingHttpServletRequest request) {
-        return Arrays.asList(request.getRequestPathInfo().getSelectors()).contains(INLINE_SELECTOR);
-    }
 
-    private byte[] getByteArray(Asset asset) throws IOException {
-        byte[] bytes = new byte[]{};
-        Resource resource = asset.getOriginal();
-        if (resource != null) {
-            InputStream stream = resource.adaptTo(InputStream.class);
-            if (stream != null) {
-                try {
-                    bytes = IOUtils.toByteArray(stream);
-                } finally {
-                    stream.close();
+    private void sendResource(SlingHttpServletRequest request, SlingHttpServletResponse response, String filename,
```

Comment:
This method does a lot more than just sending the resource. Therefore I would rename it in `handleDirectUploadResource` for example.